### ISSUE TITLE
Multiple belongsTo relations to one model breaks

### ISSUE THREAD

**Issue by ralfschimmel** [postlight/lux#683]
When model A has multiple belongsTo relations to model B, specified using:

```
class AccountMaster extends Model {
    static belongsTo = {
       aCountry: {
            model: 'country',
            inverse: 'aModels'
        },
        bCountry: {
            model: 'country',
            inverse: 'bModels'
        }
   };
}
```

It will only do a LEFT OUTER JOIN for the first relation and thus never return the second in the JSON response.

**Comment by willviles** [comment]
+1. 

I've created a [gist](https://gist.github.com/willviles/001e2a998a8f0c64862ab32e7082cebb) for what sounds like the same issue.

**Comment by willviles** [comment]
Feel like I've dropped down a bit of a rabbit hole on this issue - so any pointers are appreciated!

So, I found the offending code at [controller/utils/params-to-query.js#L53-L58](https://github.com/postlight/lux/blob/master/src/packages/controller/utils/params-to-query.js#L53-L58). Using `find` returns just the first relationship per related model.

I've got `includedFields` returning all relationships by modifying to:

```javascript
includedFields = entries(includedFields).reduce((result, field) => {
  const [key] = field;
  let [, value] = field;

  relationships.forEach(([name, { model: related }]) => {
    if (key !== related.resourceName) { return result };

    if (!value.includes(related.primaryKey)) {
      value = [related.primaryKey, ...value];
    }

    if (include && value.length === 1 && include.includes(name)) {
      value = [...value, ...related.serializer.attributes];
    } else if (!include && value.length > 1) {
      value = value.slice(0, 1);
    }

    result = {
      ...result,
      [name]: value
    };

  }) || [];

  return result;

}, {});
```

However, there's then a problem with building the SQL query in [database/query/index.js#L372-L384](https://github.com/postlight/lux/blob/master/src/packages/database/query/index.js#L372-L384).

Lux builds the following SQL query. Unfortunately, this throws an `SQLITE_ERROR: ambiguous column name: images.id`.

```SQL
SELECT "users"."id"    AS "id", 
       "users"."name"  AS "name", 
       "users"."email" AS "email", 
       "images"."id"   AS "avatar.id", 
       "images"."id"   AS "coverImage.id" 
FROM   "users" 
       LEFT OUTER JOIN "images" 
                    ON "users"."avatar_id" = "images"."id" 
       LEFT OUTER JOIN "images" 
                    ON "users"."cover_image_id" = "images"."id" 
ORDER  BY users.created_at ASC, 
          users.id ASC 
LIMIT  25
```

This seems to be an SQL aliasing issue. I hacked together a solution at [database/query/index.js#L377-L384](https://github.com/postlight/lux/blob/master/src/packages/database/query/index.js#L377-L384) to builds the SQL below. This returns the correct ids for the relationships, however the actual relationships aren't then resolved by the Lux query handler.

```SQL
SELECT "users"."id"    AS "id", 
       "users"."name"  AS "name", 
       "users"."email" AS "email", 
       "users"."avatar_id"   AS "avatar.id", 
       "users"."cover_image_id"   AS "coverImage.id" 
FROM   "users" 
       LEFT OUTER JOIN "images" 
                    ON "users"."avatar_id" = "avatar.id" 
       LEFT OUTER JOIN "images" 
                    ON "users"."cover_image_id" = "coverImage.id" 
ORDER  BY users.created_at ASC, 
          users.id ASC 
LIMIT  25
```

[database/model/index.js#L681](https://github.com/postlight/lux/blob/master/src/packages/database/model/index.js#L681)
```javascript
const props = pick(attrs, ...attributeNames.concat(relationshipNames));
// => {
//  "id": 1,
//  "name": "User Name",
//  "email": "user@email",
//  "posts": [
//    <Post model resolved properly from hasMany>
//  ],
//  "avatar": 1, /*Only get the ID for belongsTo */
//  "coverImage": 2 
//}
```

I can't seem to find the cause for the belongsTo relationships not being resolved. Can @zacharygolba perhaps shed some light?

**Comment by willviles** [comment]
Check out my fork of Lux (branch `fix-683`) to test this issue:

https://github.com/willviles/lux/tree/fix-683

**Comment by ralfschimmel** [comment]
Nice, hope @zacharygolba has some spare time soon ;-)

**Comment by willviles** [comment]
Right, I've got this working, but not sure of the best implementation of the fix.

Here's the problem. To reference multiple belongs_to relationships from the same model, we need to alias the outer joins, as I've detailed above. 

Unfortunately, when you put the LEFT_OUTER_JOIN alias as dot notation (e.g. `"avatar.id"`), it gets parsed as `"avatar"."id"`, whereas in the SELECT part of the query, it's maintained as a string:

```SQL
SELECT "users"."id"    AS "id", 
       "users"."avatar_id"   AS "avatar.id", // Kept as string
FROM   "users" 
       LEFT OUTER JOIN "images" 
                    ON "users"."avatar_id" = "avatar"."id" // Dot notation parsed
```

This means the aliasing fails. Which is a bugger, because only keys with dot notation are resolved to records as shown here [database/query/runner/utils/build-results.js#L113](https://github.com/postlight/lux/blob/master/src/packages/database/query/runner/utils/build-results.js#L113) .

If I set the aliasing like so:

```SQL
SELECT "users"."id"    AS "id", 
       "users"."avatar_id"   AS "avatar_id",
FROM   "users" 
       LEFT OUTER JOIN "images" 
                    ON "users"."avatar_id" = "avatar_id"
```

And then modify the [database/query/runner/utils/build-results.js#L113](https://github.com/postlight/lux/blob/master/src/packages/database/query/runner/utils/build-results.js#L113) to check for `key.indexOf('.') >= 0 || key.indexOf('_') >= 0`, then the records are resolved as expected.

This could be a solution, because all other attribute keys are resolved to camelCase, so underscores would be unique to this scenario. It just smells a little.

**Comment by jamemackson** [comment]
Ok, I've got my project far enough to where I'm running into this now.  @willviles / @zacharygolba - is there anything I can do to help push this fix along?

**Comment by jamemackson** [comment]
@willviles & @zacharygolba - I believe I have, somewhat accidentally, found a workaround for this it is working for me to load the second belongsTo relationship (based on latest lux in master)

context wise, I've got a `products` model that has a `colorChart` and an `alternateColorChart` belongsTo relationships, both relating to the same resource called `color-chart`.  Today I got the second relationship wired up and could set the value but then on a fresh get request the `alternateColorChart` would be null in relationships which landed me back here as is the crux of this issue.  I already had `colorChart` and `alternateColorChart` in the `hasOne` array on the product serializer.  

### the work-around?  

When I added `alternateColorChartId` to the attributes array on the serializer, then the `alternateColorChart` relationship also began populating in addition to the initial `colorChart` relationship.  

I've been able to add and remove this and consistently get working/not working results, can you guys try and confirm this also works for you? 

I haven't thought thru all the downstream implications of just adding this field to the attributes but clearly it has some effect relevant to the conversations above so I thought I'd share.

**Comment by willviles** [comment]
@jamemackson - you're right! Using the example of my gist, the following serializer config does indeed work:

```javascript
// serializers/user.js
attributes = [
  ...
  'avatarId',
  'coverImageId'
];

hasOne = [
  'avatar',
  'coverImage'
];
```

This is a good **short-term workaround**, but this isn't a particularly nice pattern. It also adds superfluous `avatar-id` and `cover-image-id` attrs to the attributes payload.

Currently, my [fix-683 branch](https://github.com/willviles/lux/tree/fix-683) ensures records are correctly added to the payload from simply specifying the relationship in `hasOne`. However included data returns blank records.
